pkg/envoy/lds/builder.go

```diff
@@ -0,0 +1,594 @@
+		filters = append(filters, tcpProxyFilter)
+	}
+
+	return filters, nil
+}
+
+// HTTPConnManagerBuilder returns the HTTP Connection Manager builder instance
+func HTTPConnManagerBuilder() *httpConnManagerBuilder { //nolint: revive // unexported-return
+	return &httpConnManagerBuilder{}
+}
+
+// StatsPrefix sets the stats prefix on the builder
+func (hb *httpConnManagerBuilder) StatsPrefix(statsPrefix string) *httpConnManagerBuilder {
+	hb.statsPrefix = statsPrefix
+	return hb
+}
+
+// RouteConfigName sets the route config name on the builder
+func (hb *httpConnManagerBuilder) RouteConfigName(name string) *httpConnManagerBuilder {
+	hb.routeConfigName = name
+	return hb
+}
+
+// DefaultFilters sets the default HTTP filters on the builder
+func (hb *httpConnManagerBuilder) DefaultFilters() *httpConnManagerBuilder {
+	hb.filters = append(hb.filters,
+		// HTTP RBAC filter - required to perform HTTP based RBAC per route
+		&xds_hcm.HttpFilter{
+			Name: envoy.HTTPRBACFilterName,
+			ConfigType: &xds_hcm.HttpFilter_TypedConfig{
+				TypedConfig: &any.Any{
+					TypeUrl: envoy.HTTPRBACFilterTypeURL,
+				},
+			},
+		},
+		// HTTP local rate limit filter - required to perform local rate limiting
+		&xds_hcm.HttpFilter{
+			Name: envoy.HTTPLocalRateLimitFilterName,
+			ConfigType: &xds_hcm.HttpFilter_TypedConfig{
+				TypedConfig: protobuf.MustMarshalAny(
+					&xds_http_local_ratelimit.LocalRateLimit{
+						StatPrefix: hb.statsPrefix,
+						// Since no token bucket is defined here, the filter is disabled
+						// at the listener level. For HTTP traffic, the rate limiting
+						// config is applied at the VirtualHost/Route level.
+						// Ref: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter#using-rate-limit-descriptors-for-local-rate-limiting
+					},
+				),
+			},
+		},
+		// Router filter - required to perform HTTP connection management
+		// *Note: the router filter must always be the last filter in the list
+		&xds_hcm.HttpFilter{
+			Name: envoy.HTTPRouterFilterName,
+			ConfigType: &xds_hcm.HttpFilter_TypedConfig{
+				TypedConfig: &any.Any{
+					TypeUrl: envoy.HTTPRouterFilterTypeURL,
+				},
+			},
+		},
+	)
+
+	return hb
+}
+
+// AddFilter adds the given HttpFilter to the builder's filter list.
+// It ensures the HTTP router filter is always the last filter in the list, which
+// is a requirement in Envoy.
+func (hb *httpConnManagerBuilder) AddFilter(filter *xds_hcm.HttpFilter) *httpConnManagerBuilder {
+	if filter == nil {
+		return hb
+	}
+
+	hb.filters = append(hb.filters, filter)
+
+	routerFilterIndex := -1
+	lastIndex := len(hb.filters) - 1
+
+	for i, f := range hb.filters {
+		if f.Name == envoy.HTTPRouterFilterName {
+			routerFilterIndex = i
+			break
+		}
+	}
+
+	// Guard against the case where multiple router filters are accidentally programmed.
+	// This panics to ensure this never happens in code.
+	if routerFilterIndex == -1 && filter.Name == envoy.HTTPRouterFilterName {
```

Comment:
How would `routerFilterIndex` be equal to -1 after matching 2 filters? There's also a break? Seems like this guards against no router filters as opposed to multiple

Reply:
Good catch. Should be `!=`

Reply:
Should there be a test added to catch this?

Reply:
Added.